asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+            "headroom_mva",
+            "Network_Reference_ID",
+        ]
+    ]
+
+    # Merge with shape data
+    wpd_df = wpd.merge(wpd_shapes, left_on="Network_Reference_ID", right_on="PRIM_NRID")
+
+    wpd_df["operator"] = "WPD"
+
+    return gpd.GeoDataFrame(
+        wpd_df[
+            [
+                "Substation_Name",
+                "Firm_Capacity_of_Substation_(MVA)",
+                "Measured_Peak_Demand_(MVA)",
+                "geometry",
+                "operator",
+            ]
+        ].rename(
+            columns={
+                "Substation_Name": "id",
+                "Firm_Capacity_of_Substation_(MVA)": "firm_capacity_mva",
+                "Measured_Peak_Demand_(MVA)": "peak_demand_mva",
+                "geometry": "geo_shape",
+            }
+        ),
+        geometry="geo_shape",
+        crs="EPSG:27700",
+    ).to_crs(CRS)
+
+
+def generate_substations_df() -> gpd.GeoDataFrame:
+    """
+    Generate a combined GeoDataFrame for all grid operators' substations.
+
+    Returns:
+        GeoDataFrame with combined substation data from all operators.
+    """
+    return pd.concat(
+        [
+            generate_enw_df(),
+            generate_npg_df(),
+            generate_spen_df(),
+            generate_ssen_df(),
+            generate_ukpn_df(),
+            generate_wpd_df(),
+        ]
+    )
+
+
+def aggregate_oa_to_lsoa(oa_gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Aggregate Output Area (OA) geometries to Lower Layer Super Output Area (LSOA) level.
+
+    Args:
+        oa_gdf: GeoDataFrame containing OA-level data with LSOA codes.
+
+    Returns:
+        GeoDataFrame aggregated to LSOA level.
+    """
+    if oa_gdf.crs is None:
+        raise ValueError(
+            "Input GeoDataFrame must have a defined coordinate reference system (CRS)"
+        )
+
+    lsoa_gdf = oa_gdf.dissolve(by="LSOA21CD", as_index=False)
+    return lsoa_gdf.reset_index(drop=True)
+
+
+def distribute_substation_headroom(
+    substations_df: gpd.GeoDataFrame,
+    lsoa_gdf: gpd.GeoDataFrame,
+    households_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Distribute substation headroom to LSOAs, weighted by household count.
+
+    Args:
+        substations_df: GeoDataFrame of substations with headroom data.
+        lsoa_gdf: GeoDataFrame of LSOA boundaries.
+        households_df: DataFrame with household counts per LSOA.
+
+    Returns:
+        DataFrame with distributed headroom per LSOA.
+    """
+    # Merge LSOA boundaries with household counts
+    lsoa_with_households = lsoa_gdf.merge(households_df, on="LSOA21CD", how="left")
+
+    # Perform spatial join between LSOAs and substations
+    joined = gpd.sjoin(
+        lsoa_with_households, substations_df, how="inner", predicate="intersects"
+    )
+
+    # Calculate total households served by each substation
+    substation_total_households = (
+        joined.groupby("index_right")["household_count"].sum().reset_index()
+    )
```

Comment:
i think it could be more explicit that `index_right` is the substation IDs, maybe this col can be renamed especially as it's used again below.

also, could this double count households, e.g. if an LSOA gets joined to two substations?